File: 1801_get_number_of_backlog_orders/get_number_of_backlog_orders.cpp
/*
 * @Date: 2023-01-02
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2023-01-02
 * @FilePath: /algorithm/1801_get_number_of_backlog_orders/get_number_of_backlog_orders.cpp
 */

#include <cassert>
#include <queue>
#include <vector>

using namespace std;

class Solution {
 public:
  int getNumberOfBacklogOrders(vector<vector<int>>& orders) {
    const int MOD = 1000000007;
    priority_queue<pair<int, int>, vector<pair<int, int>>, less<pair<int, int>>> buyOrders;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> sellOrders;
    for (auto&& order : orders) {
      int price = order[0], amount = order[1], orderType = order[2];
      if (orderType == 0) {
        while (amount > 0 && !sellOrders.empty() && sellOrders.top().first <= price) {
          auto sellOrder = sellOrders.top();
          sellOrders.pop();
          int sellAmount = min(amount, sellOrder.second);
          amount -= sellAmount;
          sellOrder.second -= sellAmount;
          if (sellOrder.second > 0) {
            sellOrders.push(sellOrder);
          }
        }
        if (amount > 0) {
          buyOrders.emplace(price, amount);
        }
      } else {
        while (amount > 0 && !buyOrders.empty() && buyOrders.top().first >= price) {
          auto buyOrder = buyOrders.top();
          buyOrders.pop();
          int buyAmount = min(amount, buyOrder.second);
          amount -= buyAmount;
          buyOrder.second -= buyAmount;
          if (buyOrder.second > 0) {
            buyOrders.push(buyOrder);
          }
        }
        if (amount > 0) {
          sellOrders.emplace(price, amount);
        }
      }
    }
    int total = 0;
    while (!buyOrders.empty()) {
      total = (total + buyOrders.top().second) % MOD;
      buyOrders.pop();
    }
    while (!sellOrders.empty()) {
      total = (total + sellOrders.top().second) % MOD;
      sellOrders.pop();
    }
    return total;
  }
};

int main() {
  {
    vector<vector<int>> orders{{10, 5, 0}, {15, 2, 1}, {25, 1, 1}, {30, 4, 0}};
    int ans = 6;
    assert(Solution().getNumberOfBacklogOrders(orders) == ans);
  }

  {
    vector<vector<int>> orders{{7, 1000000000, 1}, {15, 3, 0}, {5, 999999995, 0}, {5, 1, 1}};
    int ans = 999999984;
    assert(Solution().getNumberOfBacklogOrders(orders) == ans);
  }
}

File: golang/1801_get_number_of_backlog_orders/get_number_of_backlog_orders.go
/*
 * @Date: 2023-01-02
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2023-01-02
 * @FilePath: /algorithm/1801_get_number_of_backlog_orders/get_number_of_backlog_orders.go
 */

package main

import "container/heap"

func getNumberOfBacklogOrders(orders [][]int) (ans int) {
	buyOrders, sellOrders := hp{}, hp2{}
	for _, o := range orders {
		price, amount := o[0], o[1]
		if o[2] == 0 {
			for amount > 0 && len(sellOrders) > 0 && sellOrders[0].price <= price {
				if sellOrders[0].left > amount {
					sellOrders[0].left -= amount
					amount = 0
					break
				}
				amount -= heap.Pop(&sellOrders).(pair).left
			}
			if amount > 0 {
				heap.Push(&buyOrders, pair{price, amount})
			}
		} else {
			for amount > 0 && len(buyOrders) > 0 && buyOrders[0].price >= price {
				if buyOrders[0].left > amount {
					buyOrders[0].left -= amount
					amount = 0
					break
				}
				amount -= heap.Pop(&buyOrders).(pair).left
			}
			if amount > 0 {
				heap.Push(&sellOrders, pair{price, amount})
			}
		}
	}
	for _, p := range buyOrders {
		ans += p.left
	}
	for _, p := range sellOrders {
		ans += p.left
	}
	return ans % (1e9 + 7)
}

type pair struct{ price, left int }
type hp []pair

func (h hp) Len() int            { return len(h) }
func (h hp) Less(i, j int) bool  { return h[i].price > h[j].price }
func (h hp) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *hp) Push(v interface{}) { *h = append(*h, v.(pair)) }
func (h *hp) Pop() interface{}   { a := *h; v := a[len(a)-1]; *h = a[:len(a)-1]; return v }

type hp2 []pair

func (h hp2) Len() int            { return len(h) }
func (h hp2) Less(i, j int) bool  { return h[i].price < h[j].price }
func (h hp2) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *hp2) Push(v interface{}) { *h = append(*h, v.(pair)) }
func (h *hp2) Pop() interface{}   { a := *h; v := a[len(a)-1]; *h = a[:len(a)-1]; return v }

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	{
		orders := [][]int{{10, 5, 0}, {15, 2, 1}, {25, 1, 1}, {30, 4, 0}}
		ans := 6
		assert(getNumberOfBacklogOrders(orders) == ans)
	}

	{
		orders := [][]int{{7, 1000000000, 1}, {15, 3, 0}, {5, 999999995, 0}, {5, 1, 1}}
		ans := 999999984
		assert(getNumberOfBacklogOrders(orders) == ans)
	}
}
